service/matching/matcher.go

```diff
@@ -0,0 +1,210 @@
+		if task.syncResponseCh != nil {
+			// if there is a response channel, block until resp is received
+			// and return error if the response contains error
+			err = <-task.syncResponseCh
+			return true, err
+		}
+		return false, err
+	default: // no poller waiting for tasks
+		if rsv != nil {
+			// there was a ratelimit token we consumed
+			// return it since we did not really do any work
+			rsv.Cancel()
+		}
+		return false, nil
+	}
+}
+
+// MustOffer blocks until a consumer is found to handle this task
+// Returns error only when context is canceled or the ratelimit is set to zero (allow nothing)
+// The passed in context MUST NOT have a deadline associated with it
+func (tm *TaskMatcher) MustOffer(ctx context.Context, task *internalTask) error {
+	if _, err := tm.ratelimit(ctx); err != nil {
+		return err
+	}
+	select {
+	case tm.taskC <- task:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Poll blocks until a task is found or context deadline is exceeded
+// On success, the returned task could be a query task or a regular task
+// Returns ErrNoTasks when context deadline is exceeded
+func (tm *TaskMatcher) Poll(ctx context.Context) (*internalTask, error) {
+	select {
+	case task := <-tm.taskC:
+		if task.syncResponseCh != nil {
+			tm.scope().IncCounter(metrics.PollSuccessWithSyncCounter)
+		}
+		tm.scope().IncCounter(metrics.PollSuccessCounter)
+		return task, nil
+	case task := <-tm.queryTaskC:
+		tm.scope().IncCounter(metrics.PollSuccessWithSyncCounter)
+		tm.scope().IncCounter(metrics.PollSuccessCounter)
+		return task, nil
+	case <-ctx.Done():
+		tm.scope().IncCounter(metrics.PollTimeoutCounter)
+		return nil, ErrNoTasks
+	}
+}
+
+// PollForQuery blocks until a *query* task is found or context deadline is exceeded
+// Returns ErrNoTasks when context deadline is exceeded
+func (tm *TaskMatcher) PollForQuery(ctx context.Context) (*internalTask, error) {
+	select {
+	case task := <-tm.queryTaskC:
+		tm.scope().IncCounter(metrics.PollSuccessWithSyncCounter)
+		tm.scope().IncCounter(metrics.PollSuccessCounter)
+		return task, nil
+	case <-ctx.Done():
+		tm.scope().IncCounter(metrics.PollTimeoutCounter)
+		return nil, ErrNoTasks
+	}
+}
+
+// UpdateRatelimit updates the task dispatch rate
+func (tm *TaskMatcher) UpdateRatelimit(rps *float64) {
+	tm.limiter.UpdateMaxDispatch(rps)
+}
+
+// Rate returns the current rate at which tasks are dispatched
+func (tm *TaskMatcher) Rate() float64 {
+	return tm.limiter.Limit()
+}
+
+func (tm *TaskMatcher) ratelimit(ctx context.Context) (*rate.Reservation, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		if err := tm.limiter.Wait(ctx); err != nil {
+			return nil, err
+		}
+		return nil, nil
+	}
+
+	// how long can we wait for a ratelimit token to be available in future
+	timeout := time.Duration(common.MinInt64(int64(deadline.Sub(time.Now())), maxRateLimitWaitTime))
```

Comment:
Can you explain more about the idea?
Looks like if no deadline, which means deadline can be forever, it will wait forever (limiter.Wait(ctx) )
But if deadline is set, it won't only wait for up to maxRateLimitWaitTime. 

Why not being consistent that always honor the deadline, or both using upper bound maxRateLimitWaitTime?

Reply:
I don't totally like this part of the code either - its needed to get desired existing production behavior. Basic idea is this 

- when context deadline is set, the call comes from history - which cannot block forever - we try to block sometime to sync match - but after that, we simply write the task to db
- when context deadline is not set, the call comes from taskReader (reads backlog from db and tries to dispatch to poller) - in this case, enforcing a shorter timeout doesn't really make sense. The caller wants to block infinitely until a token can be obtained.   

How about this - I will move the part that enforces maxRateLimitWaitTime to the caller instead of being done here ? That way, the code here becomes super simple - it simply uses the context as it is provided.

Reply:
yes that make sense and the code will be much easier to understand and debug